Tasks/FtpUpload/ftpuploadtask.ts

```diff
@@ -0,0 +1,254 @@
+            path.join(remotePath, file.substring(rootFolder.length));
+        tl.debug('remoteFile: ' + remoteFile);
+
+        var stats = tl.stats(file);
+        //ensure directory is created
+        if (stats.isDirectory()) {
+            if(!flatten){
+                createRemoteDirectory(remoteFile);
+            } else {
+                tl.debug('skipping remote directory: ' + remoteFile);
+                directoriesSkipped++;
+                checkDone();
+            }
+        }
+        if (stats.isFile()) { // upload files
+            if (overwrite) {
+                uploadFile(file, remoteFile);
+            } else {
+                //todo optimize directory reading so it is only done once
+                var remoteDirname = path.normalize(path.dirname(remoteFile));
+                var remoteBasename = path.basename(remoteFile);
+                c.list(remoteDirname, function (err, list) {
+                    for (var remote of list) {
+                        if (remote.name == remoteBasename) {
+                            tl.debug('skipping file: ' + file + ' remote: ' + remoteFile + ' because it already exists');
+                            filesSkipped++;
+                            checkDone();
+                            return;
+                        }
+                    }
+                    uploadFile(file, remoteFile);
+                });
+            }
+        }
+    });
+}
+
+function createRemoteDirectory(remoteDirectory: string) {
+    tl.debug('creating remote directory: ' + remoteDirectory);
+    c.mkdir(remoteDirectory, true, function (err) {
+        if (err) {
+            c.end();
+            failTask('Unable to create remote directory: ' + remoteDirectory + ' due to error: ' + err);
+        }
+        tl.debug('remote directory successfully created: ' + remoteDirectory);
+        if(!flatten){ //
+            directoriesCreated++;
+        }
+        checkDone();
+    });
+}
+
+function uploadFile(file: string, remoteFile) {
+    tl.debug('uploading file: ' + file + ' remote: ' + remoteFile);
+    c.put(file, remoteFile, function (err) {
+        if (err) {
+            c.end();
+            failTask('upload failed: ' + remoteFile + ' due to error: ' + err);
+        } else {
+            tl.debug('successfully uploaded: ' + remoteFile);
+            filesUploaded++;
+            checkDone();
+        }
+    });
+}
+
+c.on('ready', function () {
+    if (clean) {
+        tl.debug('cleaning remote: ' + remotePath);
+        c.rmdir(remotePath, true, function (err) {
+            if (err) {
+                c.destroy();
+                failTask('Unable to clean remote folder: ' + remotePath + ' error: ' + err);
+            }
+            uploadFiles();
+        });
+    } else {
+        tl.debug('skipping clean: ' + remotePath);
+        uploadFiles();
+    }
+});
+
+var secure = serverEndpointUrl.protocol == 'ftps:' ? true : false;
```

Comment:
This needs to be a case-insensitive comparison to 'ftps:' because schemes are case-insensitive (http://tools.ietf.org/html/rfc3986#section-3.1).